super-cool-game/Assets/Scripts/Building/BuildingController.cs

```diff
@@ -0,0 +1,141 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+// Each kind of block has a numerical id
+// that is used to keep track of disposition
+// of the block in the grid that is shared between
+// client and server.
+// <summary>
+// Return a numerical value for each kind of block.
+// </summary>
+enum BlockType {
+    Spawn = 0,
+    Heal = 1
+}
+
+public class BuildingController : MonoBehaviour {
+    // Get ground to get map size
+    public GameObject ground;
+
+    // Set map max height
+    public int maxMapHeight = 40;
+
+    // Building distance from center of the map expressed in blocks
+    public int buildingDistance = 2;
+
+    // Max building width
+    public int maxBuildingWidth = 3;
+
+    // Set block width
+    public static int blockWidth = 4;
+
+    // Debug Tools
+    public bool showDebugGrid = false;
+
+    // Assign gameObjects in EDITOR
+    // to be used in code.
+    public GameObject[] blocks;
+
+    // HIDDEN IN INSPECTOR
+    // Set map width
+    [HideInInspector] public float mapWidth;
+
+    // Coordinates of the center of the map, on the top surface of it.
+    [HideInInspector] public Vector2 mapCenter;
+
+    // Reference grid of block gameobjects
+    // this is DATA
+    [HideInInspector] public List<Block> placedBlocks = new List<Block>();
+
+    // Container for blocks in world space, in order to have relative positioning
+    // this is GAMEOBJECTS
+    [HideInInspector] public GameObject grid;
+
+    void Start() {
+        Vector2 groundSize = ground.transform.GetComponent<SpriteRenderer>().bounds.size;
+
+        // Get width of the sprite rendered as ground.
+        mapWidth = groundSize.x;
+
+        // Calculate center of the map
+        mapCenter = new Vector2(ground.transform.position.x, ground.transform.position.y + (groundSize.y / 2));
+
+        // Spawn two empty gameObjects.
+        // Blocks must be children of these two.
+
+        // TODO: Find why unity spawns two gameobjects, Block Grid and Block Grid (clone).
+        grid = new GameObject();
+        grid.name = "Block Grid";
+        grid = Instantiate(grid, mapCenter, Quaternion.identity);
+    }
+
+    // <summary>
+    // Create a singleton to call functions without instancing the class directly
+    // </summary>
+    private static BuildingController _instance = null;
```

Comment:
This is ugly and unnecessary. You can easily reference the script with GetComponent on the parent gameobject